cNVMe/Driver.cpp

```diff
@@ -0,0 +1,320 @@
+				LOG_ERROR("Invalid data direction was provided");
+				pDriverCommand->DriverStatus = INVALID_DATA_DIRECTION;
+				return;
+			}
+
+			// If the data length is invalid, fail now
+			if (pDriverCommand->TransferDataSize == 0 && pDriverCommand->TransferDataDirection != NO_DATA)
+			{
+				LOG_ERROR("Transfer data size was 0 but the data direction is not no-data");
+				pDriverCommand->DriverStatus = INVALID_DATA_LENGTH;
+				return;
+			}
+
+			// If we don't have a submission queue that matches, fail now
+			auto submissionQueueItr = SubmissionQueues.find(pDriverCommand->QueueId);
+			if (submissionQueueItr == SubmissionQueues.end())
+			{
+				LOG_ERROR("Couldn't find a submission queue with the id: " + std::to_string(pDriverCommand->QueueId));
+				pDriverCommand->DriverStatus = NO_MATCHING_SUBMISSION_QUEUE;
+				return;
+			}
+
+			// If we don't know what completion queue that submission queue maps to, fail now.
+			auto pCompletionQueue = submissionQueueItr->second.getMappedQueue();
+			if (!pCompletionQueue)
+			{
+				LOG_ERROR("Couldn't find a linked completion queue for a submission queue with the id: " + std::to_string(pDriverCommand->QueueId));
+				pDriverCommand->DriverStatus = NO_LINKED_COMPLETION_QUEUE;
+				return;
+			}
+
+			// here goes nothing... send the command!
+			auto pSubmissionQueue = &submissionQueueItr->second;
+
+			// Add the CID to the command
+			pDriverCommand->Command.DWord0Breakdown.CID = getCommandIdForSubmissionQueueIdViaIncrementIfNeeded(pSubmissionQueue->getQueueId());
+
+			// create a prps object (even if we don't use it)
+			//  should stay in scope till command is done or we time out.
+			PRP prps(cnvme::Payload(pDriverCommand->TransferData, pDriverCommand->TransferDataSize), TheController.getControllerRegisters()->getMemoryPageSize());
+			if (pDriverCommand->TransferDataDirection != NO_DATA)
+			{
+				pDriverCommand->Command.DPTR.DPTR1 = prps.getPRP1();
+				pDriverCommand->Command.DPTR.DPTR2 = prps.getPRP2();
+			}
+
+			// Get a pointer to the location to place the command
+			UINT_8* theRawSubmissionQueue = MEMORY_ADDRESS_TO_8POINTER(pSubmissionQueue->getMemoryAddress());
+			NVME_COMMAND* nvmeCommand = (NVME_COMMAND*)theRawSubmissionQueue;
+			nvmeCommand += pSubmissionQueue->getTailPointer(); // Make sure we increment to the correct command.
+
+			// Copy the command into the submission queue
+			LOG_INFO("About to copy a command with an opcode of 0x" + cnvme::strings::toHexString(pDriverCommand->Command.DWord0Breakdown.OPC) + \
+				" to submission queue 0x" + cnvme::strings::toHexString(pDriverCommand->QueueId) + " and CID of 0x" + cnvme::strings::toHexString(pDriverCommand->Command.DWord0Breakdown.CID));
+
+			memcpy_s(nvmeCommand, sizeof(NVME_COMMAND), &pDriverCommand->Command, sizeof(pDriverCommand->Command));
+
+			// Move the tail pointer up and ring the doorbell.
+			pSubmissionQueue->incrementTailPointerAndRingDoorbell();
+
+			// The head pointer is not used here. Just the tail.
+
```

Comment:
Is there supposed to be code here? Maybe something that finalizes the command submission (per the "command has been sent" comment)?

Reply:
no. The idea is that the Queue class also has a head to keep track of the head when used in the controller. It isn't used here so we don't have to move it up.